src/Database/LSMTree/Internal/Chunk.hs

```diff
@@ -29,81 +35,67 @@ fromChunk (Chunk content) = content
 
--- Mutable state of a baler.
-data BalerState = BalerState
-                      [Vector Word8] -- Reverse list of queued blocks
-                      !Int           -- Total size of queued blocks
-
 {-|
-    An object that receives blocks of bytes and repackages them into chunks of
-    preferably a given maximum size.
+    An object that receives blocks of bytes and repackages them into chunks such
+    that all chunks except for a possible remnant chunk at the end are of at
+    least a given minimum size.
 -}
 data Baler s = Baler
-                   !Int                  -- Maximum chunk size in bytes
-                   !(STRef s BalerState) -- Reference to the mutable state
+                   !(MVector s Word8) -- Buffer storing queued bytes
+                   !(STRef s Int)     -- Reference to the number of queued bytes
 
 -- | Creates a new baler.
-createBaler :: Int            -- ^ Maximum chunk size in bytes
+createBaler :: Int            -- ^ Minimum chunk size in bytes
             -> ST s (Baler s) -- ^ Creation of the baler
-createBaler maxChunkSize = Baler maxChunkSize <$> (newSTRef $! BalerState [] 0)
+createBaler minChunkSize = assert (minChunkSize > 0)              $
+                           Baler                                 <$>
+                           Mutable.unsafeNew (pred minChunkSize) <*>
+                           newSTRef 0
 
 {-|
     Feeds a baler a block of bytes.
 
-    Bytes received by a bailer are generally queued for later output, but if
-    feeding new bytes makes the queued content exceed the maximum chunk size
-    then as many chunks of maximum size as possible are output.
+    Bytes received by a baler are generally queued for later output, but if
+    feeding new bytes makes the accumulated content exceed the minimum chunk
+    size then a chunk containing all the accumulated content is output.
 -}
-feedBaler :: Vector Word8 -> Baler s -> ST s [Chunk]
-feedBaler block (Baler maxChunkSize stateRef) = do
-    BalerState remnantBlocksRev remnantSize <- readSTRef stateRef
+feedBaler :: Vector Word8 -> Baler s -> ST s (Maybe Chunk)
+feedBaler block (Baler buffer remnantSizeRef) = do
+    remnantSize <- readSTRef remnantSizeRef
     let
 
         totalSize :: Int
         !totalSize = remnantSize + length block
 
-    if totalSize < maxChunkSize
+    if totalSize <= Mutable.length buffer
         then do
-                 let
-
-                     state' :: BalerState
-                     !state' = BalerState (block : remnantBlocksRev) totalSize
+                 copy (Mutable.slice remnantSize (length block) buffer) block
+                 writeSTRef remnantSizeRef totalSize
+                 return Nothing
+        else do
+                 protoChunk <- Mutable.unsafeNew totalSize
+                 Mutable.copy (Mutable.take remnantSize protoChunk)
+                              (Mutable.take remnantSize buffer)
+                 copy (Mutable.drop remnantSize protoChunk) block
```

Comment:
`unsafeCopy`

Reply:
Fixed in b9c3c88968b37a78bdf3beba6338225f2d12bcfd.